### ISSUE TITLE
False positive from `no-unused-expressions` when using nullish coalescing operator with array push

### ISSUE THREAD

**Issue by johnbillion** [eslint/eslint#13666]
**Tell us about your environment**

Node version: v14.3.0
npm version: v6.14.4
Local ESLint version: v7.8.1 (Currently used)
Global ESLint version: Not found

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?**

`@typescript-eslint/parser`

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
  "env": {
    "browser": false,
    "es2021": true
  },
  "extends": [
    "airbnb-base"
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2020,
    "sourceType": "module"
  },
  "plugins": [
    "@typescript-eslint"
  ]
}

```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

The following line gets flagged as a `no-unused-expressions` violation, despite it calling the `push()` method on an array.

```js
// This line causes a violation:
test && myarray.push(test);

// This line is fine:
myarray.push(test);
```

```bash
eslint -c .eslintrc.json --ext .ts src
```

**What did you expect to happen?**

I did not expect this line to cause a linting violation. The nullish coalescing operator is being used to check a value and push it onto an array, therefore this is not an unused expression.

**What actually happened? Please include the actual, raw output from ESLint.**

```
330:5   error  Expected an assignment or function call and instead saw an expression
no-unused-expressions
```

**Are you willing to submit a pull request to fix this bug?**

No, sorry. I'm not familiar with the ESLint internals.

**Comment by anikethsaha** [comment]
Thanks for the report. 

Can you try with `allowShortCircuit : true` [demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW51c2VkLWV4cHJlc3Npb25zOiBbMiwgeyBhbGxvd1Nob3J0Q2lyY3VpdDogdHJ1ZSB9XSovXG5cbi8vIFRoaXMgbGluZSBjYXVzZXMgYSB2aW9sYXRpb246XG50ZXN0ICYmIG15YXJyYXkucHVzaCh0ZXN0KTtcblxuLy8gVGhpcyBsaW5lIGlzIGZpbmU6XG5teWFycmF5LnB1c2godGVzdCk7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo2LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7ImpzeCI6dHJ1ZX19LCJydWxlcyI6e30sImVudiI6e319fQ==)

`airbnb-base` config I think uses `false` for `allowShortCircuit`

**Comment by johnbillion** [comment]
That does seem to be the case, thanks.

I wonder if the error message in this situation could be clarified? The message is not technically correct, even if the presence of an error is expected.
